src/Data/Vector/Mutable/Linear.hs

```diff
@@ -87,76 +97,59 @@ constant size x f
   | size < 0 =
       (error ("Trying to construct a vector of size " ++ show size) :: x #-> x)
       (f undefined)
-  | otherwise = f (Vec (size, size) (Unsafe.newMutArr size x))
+  | otherwise = Array.alloc size x (f . fromArray)
 
 -- | Allocator from a list
 fromList :: HasCallStack => [a] -> (Vector a #-> Unrestricted b) #-> Unrestricted b
-fromList xs (f :: Vector a #-> Unrestricted b) =
-  constant
-    (Prelude.length xs)
-    (error "invariant violation: uninitialized vector position")
-    (\vec -> f (build' vec))
-  where
-    build' :: Vector a #-> Vector a
-    build' = doWrites (zip xs [0..])
-
-    doWrites :: [(a,Int)] -> Vector a #-> Vector a
-    doWrites [] vec = vec
-    doWrites ((a,ix):ws) vec = doWrites ws (write vec ix a)
+fromList xs f = Array.fromList xs (f . fromArray)
 
 -- | Length of the vector
 length :: Vector a #-> (Vector a, Int)
-length = Unsafe.toLinear unsafeLength
-  where
-    unsafeLength :: Vector a -> (Vector a, Int)
-    unsafeLength v@(Vec (len, _) _) = (v, len)
+length (Vec len arr) = (Vec len arr, len)
 
 -- | Insert at the end of the vector
 snoc :: HasCallStack => Vector a #-> a -> Vector a
-snoc (Vec (len, size) ma) x
-  | len < size = write (Vec (len + 1, size) ma) len x
-  | otherwise = write (unsafeResize (max size 1 * 2) (Vec (len + 1, size) ma)) len x
+snoc (Vec size arr) x =
```

Comment:
Please rename `size` to `len` everywhere. Size should be for allocated space and length for semantically used space.

Reply:
We decided to go with the terms "size" and "capacity" following C++ reference.